### PR TITLE
[6411] New bookmark

### PR DESCRIPTION
**Pull request by KarolJagodzinski**
**Ticket**: [BAU-6411](https://bugzilla.stxnext.pl/show_bug.cgi?id=6411)

Added new bookmark(top 5 workers monthly by monthl mean time)
Added tests
Modified CSS for new bookmark

### CODE REVIEW

File: src/presence_analyzer/static/js/utils.js [hunk 1]
```diff
@@ -6,6 +6,14 @@ function parseInterval(value) {
 
 function getAvatar(user_id) {
     $.getJSON('/api/v1/user_image/'+user_id, function(img_url) {
-        $('#avatar').html('<img src="'+ img_url +'" width=100 height=100/>');
+        $('#avatar').html('<img src="'+ img_url +'" />');
+        $('#avatar').show();
+    });
+}
+
+function getAvatars(user_id) {
```
Comment (mjelonek92) [inline src/presence_analyzer/static/js/utils.js]: Why a second function?

File: src/presence_analyzer/templates/base.html [hunk 2]
```diff
@@ -47,11 +48,16 @@
                 <select id="user_id" style="display: none">
                     <option value="">--</option>
                 </select>
-                <div id="avatar">
+                <select id="month" style="display: none">
```
Comment (mjelonek92) [inline src/presence_analyzer/templates/base.html]: This should be only in the template that uses it as it's not part of the others.

File: src/presence_analyzer/templates/base.html [hunk 3]
```diff
@@ -47,11 +48,16 @@
...
+                    <option value="">--</option>
+                </select>
+                <div id="avatar" style="display: none">
+                </div>
+                <div id="avatars" style="display: none">
```
Comment (mjelonek92) [inline src/presence_analyzer/templates/base.html]: Why `avatar` and `avatars`?

File: src/presence_analyzer/tests.py [hunk 4]
```diff
@@ -178,14 +212,31 @@ def test_user_image_view_404(self):
         """
         Test api returning user's image for unexisting user.
         """
-        resp = self.client.get('/api/v1/user_image/10')
+        resp = self.client.get('/api/v1/user_image/999')
         self.assertEqual(resp.status_code, 404)
 
-    def test_presence_startend_view_404(self):
+    def test_presence_top_5_users_monthly_view(self):
+        # pylint: disable=invalid-name
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: What's the invalid name here?

File: src/presence_analyzer/utils.py [hunk 5]
```diff
@@ -151,11 +152,24 @@ def group_by_weekday(items):
     return result
 
 
-def seconds_since_midnight(time):
+def mean_by_month(items):
+    """
+    Groups mean presence by month.
+    """
+    result = [[] for month in range(12)]  # pylint: disable=unused-variable
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: `[[] for _ in range(12)]`

File: src/presence_analyzer/utils.py [hunk 6]
```diff
@@ -151,11 +152,24 @@ def group_by_weekday(items):
...
+        result[date.month-1].append(interval(start, end))
+    return [mean(intervals) for intervals in result]
+
+
+def seconds_since_midnight(time_to_calc):
     """
     Calculates amount of seconds since midnight.
     """
-    return time.hour * 3600 + time.minute * 60 + time.second
+    return time_to_calc.hour * 3600 + \
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Use parentheses instead of `\`.

File: src/presence_analyzer/views.py [hunk 7]
```diff
@@ -56,6 +56,45 @@ def users_view():
...
+@jsonify
+def months_view():
+    """
+    Months listining for dropdown.
+    """
+    return [
+        {
+            'month': month,
+            'name': calendar.month_abbr[month]
+        } for month in range(1, 13)]
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: `]` on the next line.

File: src/presence_analyzer/views.py [hunk 8]
```diff
@@ -56,6 +56,45 @@ def users_view():
         } for user_id in data_xml]
 
 
+@app.route('/api/v1/months', methods=['GET'])
+@jsonify
+def months_view():
+    """
+    Months listining for dropdown.
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: `s/listining/listing/`

File: src/presence_analyzer/views.py [hunk 9]
```diff
@@ -56,6 +56,45 @@ def users_view():
...
+    """
+    return [
+        {
+            'month': month,
+            'name': calendar.month_abbr[month]
+        } for month in range(1, 13)]
+
+
+@app.route('/api/v1/top5monthly/<int:month>', methods=['GET'])
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: `top5monthly/january` would make more sense than `top5monthly/1` in my opinion.

File: src/presence_analyzer/static/js/utils.js (hunk 1 above)
Comment (KarolJagodzinski) [inline src/presence_analyzer/static/js/utils.js]: In my new bookmark I got 5 images for once, so I needed another div called avatars for format it in CSS file.

File: src/presence_analyzer/tests.py [hunk 10]
```diff
@@ -323,6 +374,17 @@ def test_mean_time_of_presence(self):
...
         )
 
+    def test_mean_by_month(self):
+        """
+        Test grouping mean time at work by month.
+        """
+        data = utils.get_data()
+        sample_data_user = data[10]
+        result = utils.mean_by_month(sample_data_user)
+        self.assertEqual(type(result), list)
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: Can you make a full list assertion please?

File: src/presence_analyzer/tests.py [hunk 11]
```diff
@@ -67,12 +67,39 @@ def test_api_users(self):
...
+    def test_month_dropdown(self):
+        """
+        Test months listing.
+        """
+        resp = self.client.get('/api/v1/months')
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        data = json.loads(resp.data)
+        self.assertEqual(len(data), 12)
+        self.assertEqual(
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: `assertListEqual`

File: src/presence_analyzer/tests.py [hunk 12]
```diff
@@ -67,12 +67,39 @@ def test_api_users(self):
...
 
+    def test_month_dropdown(self):
+        """
+        Test months listing.
+        """
+        resp = self.client.get('/api/v1/months')
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        data = json.loads(resp.data)
+        self.assertEqual(len(data), 12)
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: Not needed with full assertions.

File: src/presence_analyzer/static/js/top_5_monthly_workers.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+            if (selected_month) {
+                loading.show();
+                chart_div.hide();
+                $('#avatars').empty()
+                $.getJSON('/api/v1/top5monthly/'+selected_month, function(result) {
+                    var finalresult =[];
+                    $.each(result, function(index, value) {
+                        finalresult.push([
+                            value['name'],
+                            parseInterval(32400),
```
Comment (mjelonek92) [inline src/presence_analyzer/static/js/top_5_monthly_workers.js]: What's `32400`?

File: src/presence_analyzer/templates/base.html (hunk 3 above)
Comment (KarolJagodzinski) [inline src/presence_analyzer/templates/base.html]: As I wrote above I needed it for set properties in CSS file.

File: src/presence_analyzer/static/js/top_5_monthly_workers.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+            var selected_month = $('#month').val(),
+                chart_div = $('#chart_div'),
+                no_data = $('#no_data').html('<b>Data not found</b>');
+            no_data.hide();
+            if (selected_month) {
+                loading.show();
+                chart_div.hide();
+                $('#avatars').empty()
+                $.getJSON('/api/v1/top5monthly/'+selected_month, function(result) {
+                    var finalresult =[];
```
Comment (mjelonek92) [inline src/presence_analyzer/static/js/top_5_monthly_workers.js]: Spaces around `=`.

File: src/presence_analyzer/templates/base.html (hunk 3 above)
Comment (mjelonek92) [inline src/presence_analyzer/templates/base.html]: Should be only in the template that uses it. Why not re-use your `avatar` one anyway?

File: src/presence_analyzer/tests.py (hunk 4 above)
Comment (KarolJagodzinski) [inline src/presence_analyzer/tests.py]: [a-z_][a-z0-9_]{2,30}$
 That's the regex for name. My test's name is just longer

File: src/presence_analyzer/static/js/top_5_monthly_workers.js (hunk 13 above)
Comment (KarolJagodzinski) [inline src/presence_analyzer/static/js/top_5_monthly_workers.js]: 32400 is a representation of time. Needed to set 9:00 as a 'start' value of timeline

Comment (mjelonek92) [inline src/presence_analyzer/static/js/top_5_monthly_workers.js]: Please add a comment explaining that. Why 9:00 is the start and not 10:00 or 5:00? Any reason?

File: src/presence_analyzer/static/js/top_5_monthly_workers.js [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+            if (selected_month) {
+                loading.show();
+                chart_div.hide();
+                $('#avatars').empty()
+                $.getJSON('/api/v1/top5monthly/'+selected_month, function(result) {
+                    var finalresult = [];
+                    $.each(result, function(index, value) {
+                        finalresult.push([
+                            value['name'],
+                            parseInterval(32400), /* set time to 9:00am as a default
```
Comment (dpolcyn) [inline src/presence_analyzer/static/js/top_5_monthly_workers.js]: /**
 *
 */

**Pull request merged**